### ISSUE TITLE
New CPU Profile Format

### ISSUE THREAD

**Issue by thlorenz** [thlorenz/flamegraph#21]
When saving a cpu profile from DevTools the emitted format is different from what this app expects (it changed).

At this point I have no bandwidth fixing this, but am documenting the changes as far as I understand them by now.

### Properties

The properties of a profile changed.

```
pad > oldKeys
[ 'typeId',
  'uid',
  'title',
  'head',
  'startTime',
  'endTime',
  'samples' ]
pad > newKeys
[ 'nodes', 'startTime', 'endTime', 'samples', 'timeDeltas' ]
```

### Nodes/Callgraph

`head` used to be the entry point to all nodes by following the nested children (which also mirrored the callgraph).

Now we have a` nodes` Array each of which has a flat Array of children ids. The actual child node needs to be resolved from the same nodes array.
The call graph needs to be reconstructed from that information as well as it now isn't mirrored like it was before.

### Samples/Time Deltas

Samples didn't change, however we have an additional Array, `timeDeltas`. This tells us the time in between each sampling tick. Turns out that time is not perfectly constant (can very by about 30%) and thus this information can be used to correct the amount of time spent in a function.
Especially when aggregating this data to create a flamegraph this correction could prove important.

## Changes Needed

The processor needs to be able to handle this new format. To just get it to work it mainly needs to be able to lookup the children by id and somehow reconstruct the callgraph.
Since these scripts were adapted from perl scripts and steps are not perfectly isolated, this change may affect multiple parts of the processor.

## Compatibility

In order to still work for the old format, the processor would need to be able to handle both, or we need a function that transforms old profiles to the new format (IMO the better solution).

**Comment by pmuellr** [comment]
I have some converters from the "old" v8-profiler format, to the new inspector "1.2" format, here: https://github.com/moar-things/moar-profile-viewer/tree/master/lib/profile

**Comment by thlorenz** [comment]
Well we'd need the opposite of that. However I believe it's better to make the processor handle the new format.
However if we had that then maybe we can drop support for the old format and use your converter to adapt it to the new format instead before processing.

**Comment by cmdcolin** [comment]
Any progress on this by chance? Really love using this tool :)

**Comment by thlorenz** [comment]
No, I'm debating to retire this tool and tell people to use [0x](https://github.com/davidmarkclements/0x) instead.
It cannot load devtools profiles either, but is much easier to use (actually lots of ideas + modules came from here) :)
